src/common/remotepermissions.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) by Olivier Goffart <ogoffart@woboq.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#pragma once
+
+#include <QString>
+#include <QMetaType>
+#include "ocsynclib.h"
+
+namespace OCC {
+
+/**
+ * Class that store in a memory efficient way the remote permission
+ */
+class OCSYNC_EXPORT RemotePermissions
+{
+private:
+    // The first bit tells if the value is set or not
+    // The remaining bits correspond to know if the value is set
+    quint16 _value = 0;
+
+    template <typename Char> // can be 'char' or 'ushort' if conversion from QString
+    void fromArray(const Char *p);
+
+public:
+    enum Permissions {
+        CanWrite = 1,             // W
+        CanDelete = 2,            // D
+        CanRename = 3,            // N
+        CanMove = 4,              // V
+        CanAddFile = 5,           // C
+        CanAddSubDirectories = 6, // K
+        CanReshare = 7,           // R
+        // Note: on the server, this means SharedWithMe, but in discoveryphase.cpp we also set
+        // this permission when the server reports the any "share-types"
+        IsShared = 8,             // S
+        IsMounted = 9,            // M
+        IsMountedSub = 10,        // m (internal: set if the parent dir has IsMounted)
+        PermissionsCount
+    };
+    RemotePermissions() = default;
+    explicit RemotePermissions(const char *);
+    explicit RemotePermissions(const QString &);
+
+    QByteArray toString();
+    bool hasPermission(Permissions p) const
+    {
+        if (isNull())
+            return p < IsShared; // when the permissions are unset, everything is allowed
+        return _value & (1 << static_cast<int>(p));
+    }
+    void setPermission(Permissions p)
+    {
+        _value |= (1 << static_cast<int>(p)) | 0x1;
+    }
+    bool isNull() const { return !(_value & 0x1); }
```

Comment:
Nit: It took me a minute to understand what `0x1` means, although I'm to blame since I only saw your comment at the declaration afterward. Maybe there could be some kind of explicit `IsNotNull = 0` value in the enum, just an idea.